### PR TITLE
Bug 1250946 – Simple Today Widget

### PR DESCRIPTION
**Pull request by jhugman**
https://bugzilla.mozilla.org/show_bug.cgi?id=1250946

This is the initial work for the Today widget: the New Tab and New Private Tab buttons. 

It includes work on the external URL scheme processing (i.e. firefox://).

### CODE REVIEW

**Comment by jhugman** [comment]
Build problems: 
```
(null): error: Couldn't load Info dictionary for <DVTFilePath:0x7f9b57fc6170:'/Users/jhugman/Library/Developer/Xcode/DerivedData/Build/Products/Fennec-iphonesimulator/Client.app/PlugIns/Today.appex'>
```

**Comment by jhugman** [comment]
Build problems fixed.

**Comment by sleroux** [comment]
I see what you mean with the 'New Private Tab' button being misaligned. Looks like there's quite a bit of aligning going on with the labels and buttons. Here's an alternative that might be simpler:

1. Instead of setting the `buttonContainer` edges to the view, don't set any constraints except for center equal to view. Instead, allow the buttons inside to 'push out' the size needed. This will guarantee the buttons be centered.

2. Not sure if it's needed but I don't think you need the `preferredContentSize` set. If the container and buttons get laid out correctly it should calculate the space it needs.

Some other things I would check out:

1. Don't think you need to set the view to be an effect view. When I remove it nothing changes. Not sure if it's just caching incorrectly or not.

2. Instead of having methods for creating the buttons I would just move these to lazy properties. That way they will be created on first access and cleans it up a bit. I would also do the same with the label method.

File: Extensions/Today/TodayViewController.swift [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        let effectView = UIVisualEffectView(effect: UIVibrancyEffect.notificationCenterVibrancyEffect())
+        effectView.frame = self.view.bounds
+        effectView.autoresizingMask = self.view.autoresizingMask
+
+        let ogView = self.view
+        self.view = effectView
+        self.view.addSubview(ogView)
+        self.view.tintColor = UIColor.clearColor()
+
+//        view.backgroundColor = UIColor.blueColor()
```
Comment (fluffyemily) [inline Extensions/Today/TodayViewController.swift]: can we remove commented code please?

File: Client/Application/AppDelegate.swift [hunk 2]
```diff
@@ -218,43 +218,50 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
-            var isPrivate: Bool = false
-            for item in (components.queryItems ?? []) as [NSURLQueryItem] {
-                switch item.name {
-                case "url":
-                    url = item.value
-                case "private":
-                    isPrivate = (item.value != "false" && item.value != "0")
-                default: ()
-                }
+        guard let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) else {
```
Comment (fluffyemily) [inline Client/Application/AppDelegate.swift]: we combine this into one statement

`guard let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme != "firefox" && components.scheme != "firefox-x-callback" else { return false } `

File: Client/Application/AppDelegate.swift [hunk 3]
```diff
@@ -218,43 +218,50 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+            return false
+        }
+        var url: String?
+        var isPrivate: Bool = false
+        for item in (components.queryItems ?? []) as [NSURLQueryItem] {
+            switch item.name {
+            case "url":
+                url = item.value
+            case "private":
+                isPrivate = (item.value != "false" && item.value != "0")
```
Comment (fluffyemily) [inline Client/Application/AppDelegate.swift]: this is where I miss NSString which has a boolValue method. I guess you could convert to NSString and then compare there. I couldn't test it as I couldn't get this to build under Xcode 7.3 (stupid compiler meltdown, nothing to do with your code)

`isPrivate = NSString(item.value).boolValue()`

File: Client/Application/AppDelegate.swift [hunk 4]
```diff
@@ -294,15 +301,12 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+        // Check if we have a URL from an external app or extension waiting to launch,
+        // then launch it on the main thread.
+        if let params = openInFirefoxParams {
+            openInFirefoxParams = nil
             dispatch_async(dispatch_get_main_queue()) {
-                if #available(iOS 9, *) {
-                    self.browserViewController.switchToPrivacyMode(isPrivate: false)
-                }
-                self.browserViewController.switchToTabForURLOrOpen(url)
+                self.launchFromURL(params)
```
Comment (fluffyemily) [inline Client/Application/AppDelegate.swift]: much cleaner, nice

**Pull request merged**